parlai/mturk/webapp/dev/app.jsx

```diff
@@ -1525,6 +1683,265 @@ class WorkerListPanel extends React.Component {
+
+    socket.onopen = () => {
+      this.setState({connected: true});
+    };
+
+    socket.onerror = socket.onclose = () => {
+      this.setState({connected: false}, function () {
+        this._socket = null;
+      });
+    };
+
+    this._socket = socket;
+  }
+
+  startTask() {
+    // Send a launch task request to the server, unpack the resulting
+    // task config and pull the custom frontend for the task.
+    this.setState({submitting: true});
+    postData('/run_task/' + this.props.task_id)
+      .then(res => res.json())
+      .then(
+        (result) => {
+          this.handleNewData(result);
+          try {
+            import(
+              /* webpackMode: "eager" */
+              `./task_components/${this.props.task_id}/components/custom.jsx`
+            ).then((custom) => {
+              setCustomComponents(custom.default);
+              if (result.task_config.frame_height === undefined) {
+                result.task_config.frame_height = 650;
+              }
+              this.setState({
+                task_loading: false, task_config: result.task_config});
+            });
+          } catch (err) {
+            // Custom react module not found
+            if (result.task_config.frame_height === undefined) {
+              result.task_config.frame_height = 650;
+            }
+            this.setState({
+              task_loading: false, task_config: result.task_config});
+          }
+
+        },
+        (error) => {
+          this.setState({
+            task_loading: false,
+            error: error,
+          });
+          console.log(error);
+          window.alert('Starting demo task failed. Error logged to console');
+        }
+      );
+  }
+
+  handleNewData(result) {
+    // Unpack data from an array of the return value of
+    // MockTurkAgent.get_update_packet()
+    let worker_names = result.data.map((w) => w.worker_id);
+    let curr_worker_data = this.state.worker_data;
+    result.data.map((w) => {
+      if (curr_worker_data[w.worker_id] === undefined) {
+        curr_worker_data[w.worker_id] = {
+          task_done: false,
+          done_text: null,
+          chat_state: 'waiting',
+          messages: [],
+          agent_id: null,
+          context: {},
+          world_state: null,
+          worker_id: w.worker_id,
+        };
+      }
+      let chat_state = 'waiting';
+      if (w.task_done) {
+        chat_state = 'done';
+      } else if (w.wants_message) {
+        chat_state = 'text_input';
+      }
+      let curr_worker = curr_worker_data[w.worker_id];
+      curr_worker.messages = curr_worker.messages.concat(w.new_messages);
+      curr_worker.task_done = w.task_done;
+      curr_worker.done_text = w.done_text;
+      curr_worker.world_state = w.status;
+      curr_worker.chat_state = chat_state;
+      curr_worker.agent_id = w.agent_id;
+      if (w.all_messages.length > curr_worker.messages.length) {
+        curr_worker.messages = w.all_messages;
```

Comment:
can you explain what this is doing here?

Reply:
Added comment. Basic idea is that if there are more messages in the thread than we think there are on the frontend, we sync to the full thread. We only choose to not use it all the time because sending messages wouldn't be able to be added to the thread until the send is received, which can sometimes take a second or so and looks clunky.